Text/XHtml/Internals.hs

```diff
@@ -109,17 +134,24 @@ instance (CHANGEATTRS b) => CHANGEATTRS (a -> b) where
       changeAttrs fn f = \ arg -> changeAttrs (fn arg) f
 
 instance ADDATTRS Html where
-      (Html htmls) ! attr = Html (map addAttrs htmls)
-        where
-              addAttrs (html@(HtmlTag { markupAttrs = attrs }) )
-                            = html { markupAttrs = attrs ++ attr }
-              addAttrs html = html
+    (Html htmls) ! attr = Html (mapDlist addAttrs htmls)
+      where
+        addAttrs html =
+            case html of
+                HtmlTag { markupAttrs = attrs, .. } ->
+                    HtmlTag
+                        { markupAttrs = attrs . (++ attr)
```

Comment:
But here, we have:

```haskell
(anchor ! [href "asdf"]) ! [theclass "qwer"]
==>
(HtmlTag { markupAttrs = id } ! [href "asdf"]) ! [theclass "qwer"]
==>
HtmlTag { markupAttrs = id . (++ [href "asdf"]) } ! [theclass "qwer"]
==>
HtmlTag { markupAttrs = id . (++ [href "asdf"]) . (++ [theclass "qwer"]) }
```

When we go to render, we apply `[]` to this. GHCI makes that:

```
λ> id . (++ ["asdf"]) . (++ ["qwer"]) $ []
["qwer","asdf"]
```

Which is - hm, a bit surprising! 

```
id . (++ xs) . (++ ys) $ []
-- delete id
(++ xs) . (++ ys) $ []
-- replace `.` with `$`
(++ xs) $ (++ ys) []
-- inline operator section
(++ xs) $ [] ++ ys
-- inline operator
([] ++ ys) ++ xs
```

The `dlist` package defines [`fromList`](https://www.stackage.org/haddock/lts-20.4/dlist-1.0/src/Data.DList.Internal.html#fromList) as `UnsafeDList . (++)` and [`append`](https://www.stackage.org/haddock/lts-20.4/dlist-1.0/src/Data.DList.Internal.html#append) as `UnsafeDList $ unsafeApplyDList xs . unsafeApplyDList ys`.

```
fromList xs = UnsafeDList . (++) $ xs
fromList xs = UnsafeDList $ (++) xs
fromList xs = UnsafeDList (xs ++)
```

So, yeah, I converted `attr` incorrectly here.